File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
  const phoneMap = new Map();
  // 해시 맵에 저장
  for (const phone of phone_book) {
    phoneMap.set(phone, true);
  }
  // 해시맵에 존재하는지 확인
  for (const phone of phone_book) {
    let prefix = ""; // 접두사
    for (let i = 0; i < phone.length - 1; i++) {
      // 같은 번호가 중복해서 들어가지 않으니, 마지막까지 확인할 필요 없음
      prefix += phone[i];
      if (phoneMap.has(prefix)) return false;
    }
  }
  return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    dic = {}
    for i in phone_book:
        dic[i] = True
    for i in phone_book:
        tmp = ""
        for j in range(len(i)-1):
            tmp += i[j]
            if tmp in dic:
                print(tmp)
                return False
    else:
        return True
            